File: modulo04-criacao-de-agentes-autonomos-novo/opspilot-go/strategies/react_strategy_test.go
package strategies

import (
	"strings"
	"testing"

	"opspilot/domain"
	"opspilot/llm"
	"opspilot/tools"
)

type fakeTool struct {
	name    string
	execute func(args map[string]any) (string, error)
}

func (t *fakeTool) Name() string        { return t.name }
func (t *fakeTool) Description() string { return "" }
func (t *fakeTool) Execute(args map[string]any) (string, error) {
	return t.execute(args)
}

func TestAnswersDirectlyWithoutToolsWhenModelHasNoToolCall(t *testing.T) {
	model := llm.NewFakeChatModelScripted([]llm.ModelResponse{llm.TextResponse("Tudo operacional.")})
	strategy := NewReactStrategy(model, nil)

	result, err := strategy.Run(domain.NewStrategyRunInput("status geral?"))
	if err != nil {
		t.Fatalf("unexpected error: %v", err)
	}
	if result.Answer != "Tudo operacional." {
		t.Fatalf("unexpected answer: %q", result.Answer)
	}
	if result.Metrics.LLMCalls != 1 {
		t.Fatalf("expected 1 llm call, got %d", result.Metrics.LLMCalls)
	}
}

func TestUsesToolBeforeFinalAnswer(t *testing.T) {
	checkStatus := &fakeTool{name: "check_provider_status", execute: func(args map[string]any) (string, error) {
		return "operacional", nil
	}}
	model := llm.NewFakeChatModelScripted([]llm.ModelResponse{
		{ToolCalls: []llm.ToolCall{{ToolName: "check_provider_status"}}},
		llm.TextResponse("Provedores operacionais."),
	})

	strategy := NewReactStrategy(model, []tools.Tool{checkStatus})
	result, err := strategy.Run(domain.NewStrategyRunInput("provedores estao ok?"))
	if err != nil {
		t.Fatalf("unexpected error: %v", err)
	}
	if result.Answer != "Provedores operacionais." {
		t.Fatalf("unexpected answer: %q", result.Answer)
	}
	if result.Metrics.LLMCalls != 2 {
		t.Fatalf("expected 2 llm calls, got %d", result.Metrics.LLMCalls)
	}
	found := false
	for _, event := range result.Trace {
		if strings.Contains(event.Content, "operacional") {
			found = true
		}
	}
	if !found {
		t.Fatal("expected trace to contain tool observation")
	}
}

func TestStopsAtMaxIterationsIfModelNeverAnswers(t *testing.T) {
	loopTool := &fakeTool{name: "loop", execute: func(args map[string]any) (string, error) {
		return "de novo", nil
	}}
	model := llm.NewFakeChatModelFunc(func(messages []llm.ChatMessage) (llm.ModelResponse, error) {
		return llm.ModelResponse{ToolCalls: []llm.ToolCall{{ToolName: "loop"}}}, nil
	})

	strategy := NewReactStrategyWithMaxIterations(model, []tools.Tool{loopTool}, 3)
	result, err := strategy.Run(domain.NewStrategyRunInput("nunca termina"))
	if err != nil {
		t.Fatalf("unexpected error: %v", err)
	}
	if result.Answer != "" {
		t.Fatalf("expected empty answer, got %q", result.Answer)
	}
	if result.Metrics.LLMCalls != 3 {
		t.Fatalf("expected 3 llm calls, got %d", result.Metrics.LLMCalls)
	}
}

File: modulo04-criacao-de-agentes-autonomos-novo/opspilot-java/src/test/java/com/opspilot/ReactStrategyTest.java
package com.opspilot;

import com.opspilot.domain.StrategyRunInput;
import com.opspilot.domain.StrategyResult;
import com.opspilot.llm.FakeChatModel;
import com.opspilot.llm.ModelResponse;
import com.opspilot.llm.ToolCall;
import com.opspilot.strategies.ReactStrategy;
import com.opspilot.tools.Tool;
import org.junit.jupiter.api.Test;

import java.util.List;
import java.util.Map;

import static org.assertj.core.api.Assertions.assertThat;

class ReactStrategyTest {

    @Test
    void answersDirectlyWithoutToolsWhenModelHasNoToolCall() {
        FakeChatModel model = new FakeChatModel(List.of(ModelResponse.text("Tudo operacional.")));
        ReactStrategy strategy = new ReactStrategy(model, List.of());

        StrategyResult result = strategy.run(StrategyRunInput.of("status geral?"));

        assertThat(result.answer()).isEqualTo("Tudo operacional.");
        assertThat(result.metrics().llmCalls()).isEqualTo(1);
    }

    @Test
    void usesToolBeforeFinalAnswer() {
        Tool checkStatus = new Tool() {
            public String name() {
                return "check_provider_status";
            }

            public String description() {
                return "";
            }

            public String execute(Map<String, Object> args) {
                return "operacional";
            }
        };
        FakeChatModel model = new FakeChatModel(List.of(
                new ModelResponse("", List.of(new ToolCall("check_provider_status", Map.of()))),
                ModelResponse.text("Provedores operacionais.")));

        ReactStrategy strategy = new ReactStrategy(model, List.of(checkStatus));
        StrategyResult result = strategy.run(StrategyRunInput.of("provedores estao ok?"));

        assertThat(result.answer()).isEqualTo("Provedores operacionais.");
        assertThat(result.metrics().llmCalls()).isEqualTo(2);
        assertThat(result.trace()).anyMatch(event -> event.content().contains("operacional"));
    }

    @Test
    void stopsAtMaxIterationsIfModelNeverAnswers() {
        Tool loopTool = new Tool() {
            public String name() {
                return "loop";
            }

            public String description() {
                return "";
            }

            public String execute(Map<String, Object> args) {
                return "de novo";
            }
        };
        FakeChatModel model = new FakeChatModel(messages ->
                new ModelResponse("", List.of(new ToolCall("loop", Map.of()))));

        ReactStrategy strategy = new ReactStrategy(model, List.of(loopTool), 3);
        StrategyResult result = strategy.run(StrategyRunInput.of("nunca termina"));

        assertThat(result.answer()).isEmpty();
        assertThat(result.metrics().llmCalls()).isEqualTo(3);
    }
}
